### ISSUE TITLE
WalletConnect doesn't work properly on Uniswap and Anyswap (Polygon Network)

### ISSUE THREAD

**Issue by hieronymus4826** [AlphaWallet/alpha-wallet-ios#3887]
When trying to connect to the Polygon version of Uniswap or Anyswap via WC on iOS I get this.

![image](https://user-images.githubusercontent.com/84448420/153036209-afc2f52d-9e57-48a8-a524-78b2a5ad3fcf.png)

Next, I enable Ethereum just to allow WC to connect, but the app is then locked on Ethereum mainnet. Clicking the 'switch to polygon' in the dapp doesn't do anything (press, no change in the network on Uniswap, Prompt - 'switch Metamask to Polygon' on Anyswap.

This has started after the latest update. 

On Android, everything works fine as it does allow you to pick up which chainID you are connecting using WalletConnect, before confirming.

**Comment by oa-s** [comment]
[hieronymus4826](https://github.com/hieronymus4826) Does Android support switching networks silently? I am facing issue with invalid `chain_id` (even if Uniswap was set to `poligon_chain_id` in response we receive `ethereum_chain_id`), when trying to connect Uniswap on Poligon

**Comment by oa-s** [comment]
Also maybe it happens as iOS doesn't support for some WalletConnect calls like : `wallet_scanQRCode`, `wallet_switchEthereumChain `, `wallet_addEthereumChain `. Will add them asap.

**Comment by hboon** [comment]
@oa-s are those valid WalletConnect calls?

**Comment by hboon** [comment]
I'm trying to reproduce this.

@hieronymus4826 can you help to describe your steps to reproducing https://github.com/AlphaWallet/alpha-wallet-ios/issues/3887#issue-1127518287, specifically which chains are enabled and when, in detail? I tried and seem to get the same behavior in Android, so I'm probably going through a different procedure.

@oa-s can you help to describe your steps (especially which chains are enabled and when) too? Especially how to trigger wallet_switchEthereumChain?

**Comment by oa-s** [comment]
Steps:
- make sure ethereum network is disabled in the AlphaWallet app;
- visit https://app.uniswap.org/#/swap (web page);
- tap `Connect wallet`;
- scan wallet connect qr code with the AlphaWallet app;
- observe that dapp wants to connect to ethereum network;
- enable ethereum network on the AlphaWallet app;
- scan wallet connect qr code with the AlphaWallet app;
- observe that dapp has connect with wallet connect:
- tap on Polygon network on web page https://app.uniswap.org/#/swap: 
<img width="301" alt="Screenshot 2022-02-10 at 07 57 26" src="https://user-images.githubusercontent.com/96283950/153347037-b7bd6d03-c426-4a96-abe8-eb48200b8fc3.png">

- observe wallet connect message `wallet_switchEthereumChain ` in console.

**Comment by oa-s** [comment]
on other hand `Anyswap` doesn't work properly for me, receive a lot of errors after connection:
```
WC: incomming text deserialization to JSONRPC 2.0 requests error: The data couldn’t be read because it isn’t in the correct format.
WC: ==> {"error":{"code":-32700,"message":"Parse error"},"id":null,"jsonrpc":"2.0"}
```

**Comment by hboon** [comment]
> * visit https://app.uniswap.org/#/swap (web page);
> //…
> <img alt="Screenshot 2022-02-10 at 07 57 26" width="301" src="https://user-images.githubusercontent.com/96283950/153347037-b7bd6d03-c426-4a96-abe8-eb48200b8fc3.png">
> 
> * observe wallet connect message `wallet_switchEthereumChain ` in console.

Ah thanks. This is relevant:

Pedro:

> This would depend on the wallet… walletconnect allows any JSON-RPC methods

https://canary.discord.com/channels/492410046307631105/833628540456140821/905050546904399892

So looks like we'll have to add at least `wallet_addEthereumChain` and `wallet_switchEthereumChain`. Might want to look at `WalletAddEthereumChainObject` and `WalletSwitchEthereumChainObject` and how they are used and try to reuse them?

Does v2 need this too? Or does the chains run concurrently?

**Comment by oa-s** [comment]
emm, in case if we add it it will be automatically supported by 2v, requests handling is the same for both versions

> Or does the chains run concurrently?

not clearly understood

**Comment by hboon** [comment]
@hieronymus4826 seems Android doesn't work in this scenario (which we are trying to fix for iOS in https://github.com/AlphaWallet/alpha-wallet-ios/issues/3887#issuecomment-1034551643):

1. Both Ethereum and Polygon are enabled
2. Scan QR code, connecting to Ethereum
3. In the dapp, switch to Polygon

Expected:

4. App detects/allows it (since it's enabled) and dapp updates

Observed:

4. Nothing happens, dapp stays in Ethereum

Can you help confirm this too?

**Comment by hboon** [comment]
> emm, in case if we add it it will be automatically supported by 2v, requests handling is the same for both versions
> 
> > Or does the chains run concurrently?
> 
> not clearly understood

I was thinking v2 works by requesting for 1 or more chains at the point when the session starts and there are additional functionality in v2 to request for more permission after that. And requests like signing are "tagged" with the already connected chain ID, so it doesn't work like v1 which has an active chain. Is that right?

**Comment by hboon** [comment]
Anyway, for v1 (and maybe for v2), `WalletSwitchEthereumChainObject` and `wallet_addEthereumChain` if the chain is already active should be simple, but if it is not enabled or not already added to the list of chains as a custom chain, it might be a bit more complicated (maybe we can trigger a restart of the UI).

**Comment by oa-s** [comment]
yeh, i am trying to integrate `DappRequestSwitchCustomChainCoordinator` and `DappRequestSwitchExistingChainCoordinator` for wallet connect

**Comment by hieronymus4826** [comment]
> I'm trying to reproduce this.
> 
> @hieronymus4826 can you help to describe your steps to reproducing [#3887 (comment)](https://github.com/AlphaWallet/alpha-wallet-ios/issues/3887#issue-1127518287), specifically which chains are enabled and when, in detail? I tried and seem to get the same behavior in Android, so I'm probably going through a different procedure.
> 
> @oa-s can you help to describe your steps (especially which chains are enabled and when) too? Especially how to trigger wallet_switchEthereumChain?

![ChainID](https://user-images.githubusercontent.com/84448420/153358674-f02cda84-3d64-4789-8ae3-5d3996d7c45e.jpg)

This is what you get when you use WalletConnect on Andoid. You can pick up the chain to match the selected network - on iOS for me it only connects to Ethereum (if its ticked in Active networks) or gives the error in he image in the original post)


> @hieronymus4826 seems Android doesn't work in this scenario (which we are trying to fix for iOS in [#3887 (comment)](https://github.com/AlphaWallet/alpha-wallet-ios/issues/3887#issuecomment-1034551643)):
> 
> 1. Both Ethereum and Polygon are enabled
> 2. Scan QR code, connecting to Ethereum
> 3. In the dapp, switch to Polygon
> 
> Expected:
> 
> 4. App detects/allows it (since it's enabled) and dapp updates
> 
> Observed:
> 
> 4. Nothing happens, dapp stays in Ethereum
> 
> Can you help confirm this too?

It does seem it's not working on Android as well. To be honest, when I did the original report I didn't test this specifically on Android (only the ChainID switch in the wallet).

**Comment by hboon** [comment]
@hieronymus4826 ah, ok, so there are 2 things here:

A. Android letting the user choose another chain (Polygon) despite the WalletConnect v1 [1] connection specifying Ethereum. I think Android lets the user choose from any enabled chain. Not sure this is proper behavior, but it's handy though. But if the chain selected is not available (eg. if the user had chose Gnosis), the dapp would be broken. But let's keep it that way.
B. We need to support both `wallet_switchEthereumChainObject` and `wallet_addEthereumChain` for WalletConnect (we only support this with our web3 provider). This needs to be done for both Android and iOS.

> Should I open a separate ticket on Android and work in parallel?

Don't worry about it. It was great that we can clarify here first and figure things out first. Now that we know, I'll create an issue for Android for (B).

[1] Referring specifically to WalletConnect v1 because v2 allows the dapp to specify connecting to one or more chains in the same session

**Comment by hboon** [comment]
@hieronymus4826 Created https://github.com/AlphaWallet/alpha-wallet-android/issues/2404.

Wonderful discovery BTW :)

**Comment by hieronymus4826** [comment]
Nice! Let me know if I can assist in any other way :)

**Comment by oa-s** [comment]
@hboon any suggestions how we can test `wallet_addEthereumChain` with Wallet Connect ?

**Comment by hboon** [comment]
>@hboon any suggestions how we can test `wallet_addEthereumChain` with Wallet Connect ? 

@oa-s manually simulate it :)

**Comment by oa-s** [comment]
Ok, i will try. Just want to test this case https://docs.metamask.io/guide/rpc-api.html#:~:text=in%20the%20future.-,%23,-Usage%20with%20wallet_switchEthereumChain

**Comment by hboon** [comment]
Thanks. But maybe also good to commit the code and have a button that is hidden in the WalletConnect page to trigger it, for future testing?

**Comment by oa-s** [comment]
We are not able to implement `wallet_addEthereumChain` in the app as there is an error in WalletConnectConnector from `'web3-react-walletconnect-connector'`. We we send error response (according to https://www.jsonrpc.org/specification#error_object):
`{"error":{"code":4902,"message":"{\"message\":\"Unrecognized chain ID 0x89. Try adding the chain using wallet_addEthereumChain first.\",\"code\":4902}"},"id":1644840328495632,"jsonrpc":"2.0"}`
via Wallet connect, the provider instance doesn't parse response properly, and in response we receive only message, and not hole error body

```
export async function switchToNetwork({ library, chainId }: SwitchNetworkArguments): Promise<null | void> {
  if (!library?.provider?.request) {
    return
  }
  const formattedChainId = hexStripZeros(BigNumber.from(chainId).toHexString())
  try {
    await library.provider.request({
      method: 'wallet_switchEthereumChain',
      params: [{ chainId: formattedChainId }],
    })
  } catch (error) {
    // 4902 is the error code for attempting to switch to an unrecognized chainId
    if (error.code === 4902) {
      const info = CHAIN_INFO[chainId]

      await library.provider.request({
        method: 'wallet_addEthereumChain',
        params: [
          {
            chainId: formattedChainId,
            chainName: info.label,
            rpcUrls: getRpcUrls(chainId),
            nativeCurrency: info.nativeCurrency,
            blockExplorerUrls: [info.explorer],
          },
        ],
      })
      // metamask (only known implementer) automatically switches after a network is added
      // the second call is done here because that behavior is not a part of the spec and cannot be relied upon in the future
      // metamask's behavior when switching to the current network is just to return null (a no-op)
      try {
        await library.provider.request({
          method: 'wallet_switchEthereumChain',
          params: [{ chainId: formattedChainId }],
        })
      } catch (error) {
        console.debug('Added network but could not switch chains', error)
      }
    } else {
      throw error
    }
  }
}

```
where `} catch (error) {` has error as String, and it should be like `{ code: 1, message: "err" }`

**Comment by hboon** [comment]
> We are not able to implement wallet_addEthereumChain

@oa-s I don't understand. Could you help to describe the flow where this happens? But we can commit #3898 for support of `wallet_switchEthereumChainObject` first, if it's done (then we should edit the commit message and PR title).

**Comment by oa-s** [comment]
@hboon `wallet_switchEthereumChainObject` works but not really as i expected for. When the call tries to switch chain id iand it doesn't supported by the app we  trigger failure event `return .value(.init(error: .unsupportedChain(chainId: targetChain.chainId)))` The Wallet connect module from `Uniswap` should detect this error in RPC response:
`{"error":{"code":4902,"message":"{\"message\":\"Unrecognized chain ID 0x89. Try adding the chain using wallet_addEthereumChain first.\",\"code\":4902}"},"id":1644840328495632,"jsonrpc":"2.0"}`
and detect error code `4902` but it doesn't. Fo some reasons it retrieves only `messsage` value from error response but not the `code` value, therefore it doesn't trigger `addChain` event

**Comment by oa-s** [comment]
> But we can commit https://github.com/AlphaWallet/alpha-wallet-ios/pull/3898 for support of wallet_switchEthereumChainObject first, if it's done (then we should edit the commit message and PR title).

yes, we can let me polish it a little bit
